examples/server/server.cpp

```diff
@@ -1826,38 +1797,36 @@ struct server_context {
         // apply context-shift if needed
         // TODO: simplify and improve
         for (server_slot & slot : slots) {
-            if (slot.ga_n == 1) {
-                if (slot.is_processing() && slot.n_past >= slot.n_ctx - 1) {
-                    if (!params.ctx_shift) {
-                        // this check is redundant (for good)
-                        // we should never get here, because generation should already stopped in process_token()
-                        slot.release();
-                        send_error(slot, "context shift is disabled", ERROR_TYPE_SERVER);
-                        continue;
-                    }
-
-                    // Shift context
-                    const int n_keep    = slot.params.n_keep + add_bos_token;
-                    const int n_left    = slot.n_past - n_keep;
-                    const int n_discard = slot.params.n_discard ? slot.params.n_discard : (n_left / 2);
+            if (slot.is_processing() && slot.n_past >= slot.n_ctx - 1) {
+                if (!params.ctx_shift) {
+                    // this check is redundant (for good)
+                    // we should never get here, because generation should already stopped in process_token()
+                    slot.release();
+                    send_error(slot, "context shift is disabled", ERROR_TYPE_SERVER);
+                    continue;
+                }
```

Comment:
@ngxson I think the comment is not entirely correct because in `process_token()` we check agains the training context length (`n_ctx_train`), while the slot's context `slot.n_ctx` could be smaller. What do you think?

Reply:
Hmm no, I did add a check against `slot.n_ctx`. Is this what you're looking for?

https://github.com/ggerganov/llama.cpp/blob/11ac9800aff532715a5bc7991062c68ba3472e6e/examples/server/server.cpp#L1184-L1191

Reply:
Ah, I missed that, thanks.

Shouldn't we check this actually:

```c++
    if (slot.n_prompt_tokens + slot.n_decoded >= n_ctx) {`
```

Reply:
Ah yeah I misunderstood `n_decoded`. Yeah, maybe we even need `(int) system_tokens.size() + slot.n_prompt_tokens` because `system_tokens` is already in KV cache before the first decode

Reply:
No sorry I haven't see https://github.com/ggerganov/llama.cpp/issues/9811